builtin/clean.c

```diff
@@ -171,6 +171,15 @@ static int remove_dirs(struct strbuf *path, const char *prefix, int force_flag,
 		goto out;
 	}
 
+	/* first try the simple approach, recursion may not be needed if empty dir, symlink, junction, etc. */
+	res = dry_run ? 0 : unlink(path->buf);
+	if (res == 0) {
+		printf(dry_run ?  _(msg_would_remove) : _(msg_remove), path->buf);
+		ret = res;
+		goto out;
```

Comment:
Interesting idea. I worry, however, that this breaks dry runs with real directory structures, as nothing will recurse in dry-run mode anymore, correct?

I *think* that this is going into the right direction, though, and we may need something similar for `git rm`. Maybe we can be a bit cleverer and introduce something like `is_mount_point(const char *path)` and use it here to skip the recursion?

This `is_mount_point()` I have in mind would work a little like `is_dir_sep()`: for Windows, we define it here: https://github.com/git-for-windows/git/blob/0b42e3c56de10a2252d2b7bb4c86c0096171f922/compat/mingw.h#L492-L496 For everything else, here: https://github.com/git-for-windows/git/blob/0b42e3c56de10a2252d2b7bb4c86c0096171f922/git-compat-util.h#L420-L426

Reply:
Come to think of it: Unix bind mounts are very similar in concept to junction points.

According to [this StackOverflow answer](https://unix.stackexchange.com/questions/151390/how-to-check-a-particular-directory-is-mounted), one can test whether a path is "bind-mounted" by `stat()`ing its `.` and `..` and testing whether their `st_dev` fields are different.

I think this is what the default implementation in `git-compat-util.h` should do, too.

Reply:
To make myself clear: I suggest to introduce the function `is_mount_point()` and use

```c
if (is_mount_point(path->buf)) {
[...]
```

instead of

```c
res = dry_run ? 0 : unlink(path->buf);
if (res == 0) {
[...]
```

I would also suggest to adjust the comment `/* first try the simple approach, recursion may not be needed if empty dir, symlink, junction, etc. */`:

1. `unlink()` on an empty directory is undefined, IIRC. On some systems it may succeed, on others it would fail with `EISDIR`.
2. if the path in question refers to a symbolic link, the `remove_dirs()` function is not called: https://github.com/git-for-windows/git/blob/e1848984d1004040ec5199e749b5f282ddf4bb09/builtin/clean.c#L202-L211 and https://github.com/git-for-windows/git/blob/e1848984d1004040ec5199e749b5f282ddf4bb09/builtin/clean.c#L1029-L1036 are the only call sites, and both are guarded behind `if (S_ISDIR(st.st_mode)) { ... }`.
3. if you talk about "junctions" in the context of Git for Windows, many developers will understand. The core Git people, who are almost exclusively developing on (and unfortunately also for) Linux, will not understand what you are talking about, however. They will only understand the term "mount point". Which a junction is, sort of.

And now that I saw the call sites for `remove_dirs()`, I think the correct spot to add the `is_mount_point()` condition would be just after the `S_ISDIR(st.st_mode)` condition: if it is marked as a directory *and* if it is not a mount point, then we should go on to call `remove_dirs()`. Otherwise we can fall back to `unlink()`.

But I guess that would not fly well on the Git mailing list... No idea what they would think appropriate if a directory inside a Git repository was a mount point (I don't think that any core Git developer saw such a scenario, ever)...

In any case, I think this would get us further to submit the patches to the Git mailing list, and there we could have a good discussion how to proceed. And hopefully many eyes will see it and hopefully we will come up with a satisfactory solution.

How does that sound?